src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs

```diff
@@ -81,49 +115,42 @@ internal static SafeSslHandle AllocateSslContext(SslProtocols protocols, SafeX50
                     {
                         // user explicitly asks for TLS 1.3 but their policy is not compatible with TLS 1.3
                         throw new SslException(
-                            SR.Format(SR.net_ssl_encryptionpolicy_notsupported, policy));
+                            SR.Format(SR.net_ssl_encryptionpolicy_notsupported, sslAuthenticationOptions.EncryptionPolicy));
                     }
                 }
 
-                if (CipherSuitesPolicyPal.ShouldOptOutOfLowerThanTls13(sslAuthenticationOptions.CipherSuitesPolicy, policy))
+                if (CipherSuitesPolicyPal.ShouldOptOutOfLowerThanTls13(sslAuthenticationOptions.CipherSuitesPolicy, sslAuthenticationOptions.EncryptionPolicy))
                 {
                     if (!CipherSuitesPolicyPal.WantsTls13(protocols))
                     {
                         // We cannot provide neither TLS 1.3 or non TLS 1.3, user disabled all cipher suites
                         throw new SslException(
-                            SR.Format(SR.net_ssl_encryptionpolicy_notsupported, policy));
+                            SR.Format(SR.net_ssl_encryptionpolicy_notsupported, sslAuthenticationOptions.EncryptionPolicy));
                     }
 
                     protocols = SslProtocols.Tls13;
                 }
 
                 // Configure allowed protocols. It's ok to use DangerousGetHandle here without AddRef/Release as we just
                 // create the handle, it's rooted by the using, no one else has a reference to it, etc.
-                Ssl.SetProtocolOptions(innerContext.DangerousGetHandle(), protocols);
+                Ssl.SslCtxSetProtocolOptions(innerContext, protocols);
 
-                // Sets policy and security level
-                if (!Ssl.SetEncryptionPolicy(innerContext, policy))
+                if (sslAuthenticationOptions.EncryptionPolicy != EncryptionPolicy.RequireEncryption)
                 {
-                    throw new SslException(
-                        SR.Format(SR.net_ssl_encryptionpolicy_notsupported, policy));
+                    // Sets policy and security level
+                    if (!Ssl.SetEncryptionPolicy(innerContext, sslAuthenticationOptions.EncryptionPolicy))
+                    {
+                        throw new SslException( SR.Format(SR.net_ssl_encryptionpolicy_notsupported, sslAuthenticationOptions.EncryptionPolicy));
+                    }
                 }
 
-                // The logic in SafeSslHandle.Disconnect is simple because we are doing a quiet
-                // shutdown (we aren't negotiating for session close to enable later session
-                // restoration).
-                //
-                // If you find yourself wanting to remove this line to enable bidirectional
-                // close-notify, you'll probably need to rewrite SafeSslHandle.Disconnect().
-                // https://www.openssl.org/docs/manmaster/ssl/SSL_shutdown.html
```

Comment:
I see two different places now that are calling SSL_shutdown:

https://github.com/dotnet/runtime/blob/2c275e87c0033da3830bae29059d9ee5b663cf58/src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Ssl.cs#L351-L371

and

https://github.com/dotnet/runtime/blob/0851fee685f5a849bb3e554c4765ab3830e6bb7b/src/libraries/System.Net.Security/src/System/Net/Security/SslStreamPal.Unix.cs#L216-L246

Are we doing shutdown properly?